handler/src/main/java/io/netty/handler/ssl/ApplicationProtocolNegotiationHandler.java

```diff
@@ -78,24 +77,17 @@ protected ApplicationProtocolNegotiationHandler(String fallbackProtocol) {
     }
 
     @Override
-    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
-        // FIXME: There is no way to tell if the SSL handler is placed before the negotiation handler.
-        final SslHandler sslHandler = ctx.pipeline().get(SslHandler.class);
-        if (sslHandler == null) {
-            throw new IllegalStateException(
-                    "cannot find a SslHandler in the pipeline (required for application-level protocol negotiation)");
-        }
-
-        this.sslHandler = sslHandler;
-    }
-
-    @Override
     public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
         if (evt instanceof SslHandshakeCompletionEvent) {
             ctx.pipeline().remove(this);
 
             SslHandshakeCompletionEvent handshakeEvent = (SslHandshakeCompletionEvent) evt;
             if (handshakeEvent.isSuccess()) {
+                SslHandler sslHandler = ctx.pipeline().get(SslHandler.class);
```

Comment:
LGTM but just to throw it out there... SslHandler could fire a SslHandshakeCompletionEvent with a reference to itself. It may or may not be "cheaper" than the `ctx.pipeline().get(SslHandler.class)` lookup.